[Code Completion] Split code completion results into context free part that can be cached and contextual part displayed to the user

ahoppen opened the pull request:
This is the first (of two) refactoring part of #40399, which has gotten a little big.

This allows makes the distinction between cachable and non-cachable properties cleaner and allows us to more easily compute contextual information (like type relations) for cached items later.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented on include/swift/IDE/CodeCompletion.h:
```diff
@@ -612,30 +618,420 @@ enum class CompletionKind {
+  /// thus making the pointer inside \c Ty invalid.
+  ASTContext *ContextOfTy;
+
+  /// Maps USRs to \c CodeCompletionResultTypes that have already been
+  /// constructed.
+  static llvm::StringMap<CodeCompletionResultType *> USRCache;
+
+  /// Maps Types to \c CodeCompletionResultTypes that have already been
+  /// constructed.
+  static llvm::DenseMap<Type, CodeCompletionResultType *> TypeCache;
```
Oh, I thought that cache would only be introduced by the next PR, sorry. 

I will store them in a dedicate arena with https://github.com/ahoppen/swift/commit/e9ef698fa91625f027221a403e58b691d2bfe9a8 but because that change is >200 LOC as well, I didn’t want to inflate this PR even further. Do you think we should pull that commit into this PR or should we just fix it in a follow-up PR?

We can’t store `CodeCompletionResultType` in the `ASTContext` because they are used for cached results, which are used for multiple code completion sessions and thus outlive the `ASTContext`.

rintaro replied:
> We can’t store CodeCompletionResultType in the ASTContext because they are used for cached results, which are used for multiple code completion sessions and thus outlive the ASTContext.

I'm not sure the cache should outlive the ASTContext. An ASTContext is used for multiple code completion sessions as long as fast-completion is working. If fast-completion is not working and the ASTContext is destroyed, I think the cache should be cleared, or what cache can we reuse?

ahoppen commented:
@swift-ci Please smoke test Windows

ahoppen commented:
@swift-ci Please test Windows

ahoppen commented:
I refactored the PR to make a cleaner distinction between USR-based and ASTContext-based types. For now `ContextFreeCodeCompletionResult` stores a `CodeCompletionResultType ` that is always backed by a `ASTContext`-bound `swift::Type`. That design doesn’t allow caching the types yet. 

In a follow-up PR I will add the option to back `CodeCompletionResultType` by a `swift::ide::USRBasedType` which can be cached and supports computation of convertible type relations by storing the USRs of its supertype. The current WIP of that PR lives at https://github.com/ahoppen/swift/tree/pr/compute-typerelation-for-global-cached-results.

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on include/swift/IDE/CodeCompletion.h:
```diff
@@ -612,214 +617,357 @@ enum class CompletionKind {
+/// A single code completion result enriched with information that depend on
+/// the completion's usage context.
+class CodeCompletionResult {
+  ContextFreeCodeCompletionResult ContextFree;
```
`ContextFreeCodeCompletionResult` is fairly big object. Instead of copying it, could `CodeCompletionResult` just hold a reference to it?

ahoppen replied:
I thought about the same but decided against it, because 
- this way, we are sure that `ContextFreeCodeCompletionResult` is allocated by the same allocator as `CodeCompletionResult` and we don’t run into issues that `ContextFreeCodeCompletionResult` might get deallocated while `CodeCompletionResult` is still referencing it
- we only need to copy it when constructing `CodeCompletionResult` and I don’t think it makes much of a performance difference
- We don’t have to do an indirection to `ContextFreeCodeCompletionResult` when accessing context-free properties of a `CodeCompletionResult`

Do you think this makes sense or would you think it’s worth to measure the performance impact of the copy?

ahoppen committed:
[Code Completion] Split code completion results into context free part that can be cached and contextual part displayed to the user

This allows makes the distinction between cachable and non-cachable properties cleaner and allows us to more easily compute contextual information (like type relations) for cached items later.

ahoppen committed:
[Code Completion] Extract NotRecommendedReason and CodeCompletionResultKind to top-level types

Preliminary work to split `CodeCompletionResult` into a context-free and contextual part.

ahoppen commented:
I have rebased this PR now that https://github.com/apple/swift/pull/40724 is merged and tried to make it as easy to review as possible:
- 707e6e855bfc3a6cc65c0d71b408bf63ee9b56ab is a super simple refactoring that just pull out a few nested types from `CodeCompletionResult` to the top-level
- e0253ca624e2709c3e4c5a2f88344251c0b72888 performs the actual split into contextual and context-free result bits.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test Windows

ahoppen commented:
@swift-ci Please smoke test Windows

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on include/swift/IDE/CodeCompletion.h:
```diff
@@ -630,21 +630,287 @@ enum class CodeCompletionDiagnosticSeverity : uint8_t {
+  CodeCompletionResultKind Kind : 3;
+  static_assert(int(CodeCompletionResultKind::MAX_VALUE) < 1 << 3, "");
+
+  /// Opaque kind that is interpreted differently depending on
```
Nit: Incomplete comment? Probably:
```suggestion
  /// Opaque kind that is interpreted differently depending on 'Kind'.
```

ahoppen replied:
Thanks. That comment pre-dated the use of a union here. I just removed it.

rintaro reviewed (approved):
LGTM!

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test
